### PR TITLE
feat: Poseidon2 hash function

### PR DESCRIPTION
**Pull request by danny-shterman**
- Some values of t not supported at this moment for large fields (>4B).
- Sponge hash no supported

cuda-backend-branch: danny_poseidon2/v3.2

### CODE REVIEW

File: icicle/run.log.cpu_works
Comment (LeonHibnik) [inline icicle/run.log.cpu_works:1]: Remove log files and unrelated files from the PR

File: icicle/tests/test_hash_api.cpp [hunk 1]
```diff
@@ -687,6 +688,252 @@ TEST_F(HashApiTest, MerkleTreeLarge)
...
+// p = 0x30644e72e131a029b85045b68181585d2833e84879b9709143e1f593f0000001
+
+#include "icicle/fields/field_config.h"
+// #include "icicle/hash/poseidon2_constants/constants/bn254_poseidon2.h"
+// #include "icicle/hash/poseidon2_constants/constants/bw6_761_poseidon2.h"
+#include "icicle/hash/poseidon2_constants/constants/babybear_poseidon2.h"
+
+using namespace field_config;
+// using namespace poseidon2_constants_bn254;
+// using namespace poseidon2_constants_bw6_761;
```
Comment (LeonHibnik) [inline icicle/tests/test_hash_api.cpp:701]: Remove comments if irrelevant

File: icicle/include/icicle/hash/poseidon2_constants/constants/goldilocks_poseidon2.h
Comment (LeonHibnik) [inline icicle/include/icicle/hash/poseidon2_constants/constants/goldilocks_poseidon2.h:1]: We currently don't support goldilocks in ICICLE, don't se a way to build these, can remove

Comment (danny-shterman) [reply icicle/include/icicle/hash/poseidon2_constants/constants/goldilocks_poseidon2.h:1]: Done.

File: icicle/run.log.cpu_works
Comment (danny-shterman) [reply icicle/run.log.cpu_works:1]: Done.

File: icicle/backend/cpu/src/hash/cpu_merkle_tree.cpp [hunk 2]
```diff
@@ -40,34 +40,42 @@ namespace icicle {
 
         if (0 < layer_idx) {
           // update nof_hashes to the next layer (below)
-          const uint64_t cur_layer_input_size = layer_hashes[layer_idx].default_input_chunk_size();
-          nof_hashes = nof_hashes * cur_layer_input_size / layer_hashes[layer_idx - 1].output_size();
+          const uint64_t next_layer_total_size = layer_hashes[layer_idx - 1].output_size();
```
Comment (yshekel) [inline icicle/backend/cpu/src/hash/cpu_merkle_tree.cpp:43]: did you mean to edit this file? Maybe you are not merged with main?

File: icicle/include/icicle/fields/field.h [hunk 3]
```diff
@@ -83,6 +84,27 @@ class Field
     return omega;
   }
 
+  static HOST_INLINE Field
+  hex_str2scalar(const std::string& in_str) // The input string should be in a hex format - 0xABCD....
+  {
+    assert(in_str.substr(0, 2) == "0x" && "The input string is not in hex format!");
```
Comment (yshekel) [inline icicle/include/icicle/fields/field.h:90]: note that assert is only in debug mode so consider using ICICLE_ASSERT

File: icicle/include/icicle/utils/utils.h [hunk 4]
```diff
@@ -34,6 +34,6 @@ static void print_bytes(const std::byte* data, const uint nof_elements, const ui
       std::cout << std::hex << std::setw(2) << std::setfill('0')
                 << static_cast<int>(data[element_idx * element_size + byte_idx]);
     }
-    std::cout << std::dec << ",\n";
+    std::cout << ",\n";
```
Comment (yshekel) [inline icicle/include/icicle/utils/utils.h:37]: why? this makes all future prints be in hex

File: icicle/tests/test_hash_api.cpp [hunk 5]
```diff
@@ -8,22 +8,56 @@
...
+      "%s: %.3f ms\n", msg, FpMicroseconds(std::chrono::high_resolution_clock::now() - timer##_start).count() / 1000);
+
 static bool VERBOSE = true;
 static int ITERS = 1;
+static inline std::string s_main_target;
+static inline std::string s_reference_target;
+static inline std::vector<std::string> s_registered_devices;
 
-class HashApiTest : public IcicleTestBase
+class HashApiTest : public ::testing::Test
```
Comment (yshekel) [inline icicle/tests/test_hash_api.cpp:33]: you messed up the merge!! all this code moved to test base.

File: icicle/tests/test_hash_api.cpp [hunk 6]
```diff
@@ -135,21 +169,21 @@ TEST_F(HashApiTest, KeccakLarge)
   auto output_main_case_2 = std::make_unique<std::byte[]>(output_size * config.batch);
   auto output_ref = std::make_unique<std::byte[]>(output_size * config.batch);
 
-  ICICLE_CHECK(icicle_set_device(IcicleTestBase::reference_device()));
+  ICICLE_CHECK(icicle_set_device(s_reference_target));
```
Comment (yshekel) [inline icicle/tests/test_hash_api.cpp:172]: all this file messed up the merge.

File: icicle/tests/test_hash_api.cpp [hunk 7]
```diff
@@ -878,18 +912,260 @@ TEST_F(HashApiTest, MerkleTreeLarge)
...
 
   #include "icicle/fields/field_config.h"
-  #include "icicle/hash/poseidon_constants/bn254_poseidon.h"
 
 using namespace field_config;
-using namespace poseidon_constants_bn254;
+
+  #include "icicle/hash/poseidon2.h"
+
+// // DEBUG. This test could run only with bn254 curve. Dont remove!!!
```
Comment (yshekel) [inline icicle/tests/test_hash_api.cpp:924]: so why not put under ifdef BN254?

File: icicle/include/icicle/fields/field.h (hunk 3 above)
Comment (danny-shterman) [reply icicle/include/icicle/fields/field.h:90]: No.
I didnt touch it

File: icicle/include/icicle/utils/utils.h (hunk 4 above)
Comment (danny-shterman) [reply icicle/include/icicle/utils/utils.h:37]: It's not me.....
Didnt touch it

File: icicle/tests/test_hash_api.cpp (hunk 6 above)
Comment (danny-shterman) [reply icicle/tests/test_hash_api.cpp:172]: I didnt merge it
Leon is probably taking care of that

File: icicle/backend/cpu/src/hash/cpu_merkle_tree.cpp (hunk 2 above)
Comment (danny-shterman) [reply icicle/backend/cpu/src/hash/cpu_merkle_tree.cpp:43]: No
It's not me

File: icicle/tests/test_hash_api.cpp (hunk 7 above)
Comment (danny-shterman) [reply icicle/tests/test_hash_api.cpp:924]: Leon wants tests to run for all fields/curves
This test is there only if debug is needed in the future
The functionality of this test is covered by other tests

File: icicle/backend/cpu/src/hash/cpu_merkle_tree.cpp (hunk 2 above)
Comment (yshekel) [reply icicle/backend/cpu/src/hash/cpu_merkle_tree.cpp:43]: Why did you resolve? It's an issue

**Review by LeonHibnik — APPROVED** [review]
